File: fixtures/scale/java/large/src/cryptobench/ciphers/beaufort/BeaufortCipher.java
package cryptobench.ciphers.beaufort;

import cryptobench.core.Cipher;
import cryptobench.util.Alphabet;

/** Reciprocal variant: ciphertext is key letter minus plaintext letter. */
public class BeaufortCipher implements Cipher {
    private final BeaufortKey key;

    public BeaufortCipher(BeaufortKey key) {
        this.key = key;
    }

    @Override
    public String name() {
        return "beaufort";
    }

    @Override
    public String encrypt(String plaintext) {
        return transform(Alphabet.clean(plaintext), true);
    }

    @Override
    public String decrypt(String ciphertext) {
        return transform(Alphabet.clean(ciphertext), false);
    }

    private String transform(String text, boolean forward) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < text.length(); i++) {
            int x = Alphabet.indexOf(text.charAt(i));
            int k = Alphabet.indexOf(key.keyCharAt(i));
            sb.append(Alphabet.charAt(forward ? k - x : k - x));
        }
        return sb.toString();
    }
}

File: fixtures/scale/kotlin/large/src/cryptobench/ciphers/beaufort/BeaufortCipher.kt
package cryptobench.ciphers.beaufort

import cryptobench.core.Cipher
import cryptobench.util.Alphabet

/** Reciprocal variant: ciphertext is key letter minus plaintext letter. */
class BeaufortCipher(private val key: BeaufortKey) : Cipher {

    override fun name(): String = "beaufort"

    override fun encrypt(plaintext: String): String = transform(Alphabet.clean(plaintext), true)

    override fun decrypt(ciphertext: String): String = transform(Alphabet.clean(ciphertext), false)

    private fun transform(text: String, forward: Boolean): String {
        val sb = StringBuilder()
        for (i in text.indices) {
            val x = Alphabet.indexOf(text[i])
            val k = Alphabet.indexOf(key.keyCharAt(i))
            sb.append(Alphabet.charAt(if (forward) k - x else k - x))
        }
        return sb.toString()
    }
}
